suite-native/device/src/middlewares/deviceConnectionMiddleware.ts

```diff
@@ -0,0 +1,89 @@
+import {
+    Dispatch,
+    ListenerEffectAPI,
+    UnknownAction,
+    createListenerMiddleware,
+} from '@reduxjs/toolkit';
+
+import {
+    selectIsDeviceConnectedAndAuthorized,
+    selectIsDeviceInitialized,
+    selectIsDeviceUsingPassphrase,
+} from '@suite-common/wallet-core';
+import { selectIsFirmwareInstallationRunning } from '@suite-native/firmware';
+import {
+    AuthorizeDeviceStackRoutes,
+    RootStackRoutes,
+    checkIsRouteAccessAllowed,
+    navigationContainerRef,
+} from '@suite-native/navigation';
+import { selectIsCoinEnablingInitFinished } from '@suite-native/settings';
+
+import { NativeDeviceRootState, selectIsDeviceCompromised } from '../selectors';
+import { isDeviceConnectAction } from '../utils';
+
+const DEVICE_CONNECTION_BLACKLISTED_ROUTES = [
+    RootStackRoutes.DeviceCompromisedModal,
+    RootStackRoutes.DeviceOnboardingStack,
+    RootStackRoutes.SendStack,
+];
+
+export const deviceConnectionMiddleware = createListenerMiddleware<NativeDeviceRootState>();
+
+const handleDeviceConnectNavigation = ({
+    isCoinEnablingInitFinished,
+}: {
+    isCoinEnablingInitFinished: boolean;
+}) => {
+    // If coin enabling is not finished, it takes priority over connecting screen
+    if (isCoinEnablingInitFinished) {
+        navigationContainerRef.navigate(RootStackRoutes.AuthorizeDeviceStack, {
+            screen: AuthorizeDeviceStackRoutes.ConnectingDevice,
+        });
+    } else {
+        navigationContainerRef.navigate(RootStackRoutes.CoinEnablingInit);
+    }
+};
+
+deviceConnectionMiddleware.startListening({
+    predicate: (action, currentState) =>
+        isDeviceConnectAction(action) &&
+        // TODO this should dissappear after we merge device onboarding redirect here as well
+        // https://github.com/trezor/trezor-suite/issues/20157
+        // If device is not initialized and is compromised, we display the modal (reason why this condition is here) and then want to redirect to uninitialized device landing.
+        (selectIsDeviceInitialized(currentState) || selectIsDeviceCompromised(currentState)),
+    effect: (
+        _action: UnknownAction,
+        { getState }: ListenerEffectAPI<NativeDeviceRootState, Dispatch<UnknownAction>>,
+    ) => {
+        const shouldNavigateToDeviceCompromisedModal = selectIsDeviceCompromised(getState());
+        const isCoinEnablingInitFinished = selectIsCoinEnablingInitFinished(getState());
+
+        const isDeviceConnectionNavigationAllowed = checkIsRouteAccessAllowed({
+            blacklist: DEVICE_CONNECTION_BLACKLISTED_ROUTES,
+        });
+
+        if (!isDeviceConnectionNavigationAllowed) return;
```

Comment:
I'd personally inline the variable, since it's used just once and the function being called is descriptive enough.

Reply:
👌 fixed